flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBase.java

```diff
@@ -72,6 +74,16 @@
         ClosureCleaner.clean(builder, ExecutionConfig.ClosureCleanerLevel.RECURSIVE, true);
     }
 
+    /**
+     * Set writes to be synchronous (block until writes are completed).
+     *
+     * @param timeout Maximum number of seconds to wait for write completion
+     */
+    public void setSynchronousWrites(int timeout) {
```

Comment:
This seems strange. The CassandraSinkBase explicitly contains logic to make sure all async requests have completed when shutting down or creating a checkpoint (#flush).

If some issue now appears in the test that implies that either
a) this logic is now broken, which would be problematic in general since it could cause data-loss
b) despite writes being acknowledged they aren't persisted properly, at which point I wonder what makes the synchronous call that different (maybe it just works because data is sent at a lower rate)

In any case this needs further investigations.

Reply:
To give some more details about the uncovered race condition I mentioned: when I upgraded the versions `CassandraConnectorITCase#testCassandraBatchPojoFormat` started to fail claiming no records were written. So the problem was not on `CassandraSinkBase` but on `CassandraPojoOutputFormat`. The problem was that the Cassandra session was closed before the end of the asynchronous writes leading to a Cassandra exception saying that the session is already closed. That is why I added the option for synchronous write on `CassandraPojoOutputFormat` so that `sink.writeRecord(pojo)` becomes a blocking call and `sink.close()` is not called until the write is actually done. Then I generalized this option to all sinks for coherence. So, in short, there was no problem with CassandraSinkBase and subclasses, the `flush` behavior works just fine.

Reply:
then we should instead introduce the flushing logic into the output format imo.

Reply:
Ok I can do that if you prefer. Do you want that I drop all the synchronous write features on all the sinks ?

Reply:
yes. I don't think you'd ever want to enable that in production (which is what the API should be limited to).

Reply:
:+1:

Reply:
I dropped the commits

Reply:
> I dropped the commits

@echauchot All the `sink.setSynchronousWrites` are still in. Did I got the conversation right that they all should be dropped from this PR ?

Reply:
@mzuehlke yes I have push the drop yet because I'm replacing the synchronous write by an impl of flush mechanism as part of [FLINK-27457](https://issues.apache.org/jira/browse/FLINK-27457). I'm done with prod code I just need to add the tests for the flush mechanism. I should be able to push all of that soon.